File: core/src/main/java/it/asso/core/service/CarattereService.java
package it.asso.core.service;

import it.asso.core.dao.animali.animale.CaratteriDAO;
import it.asso.core.model.animali.animale.Caratteri;
import it.asso.core.model.animali.animale.TipoCarattere;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.List;

@Service
public class CarattereService {

    private final CaratteriDAO caratteriDao;

    public CarattereService(CaratteriDAO caratteriDao) {
        this.caratteriDao = caratteriDao;
    }

    public List<Caratteri> getCaratteriByAnimale(String idAnimale) {
        List<Caratteri> list = caratteriDao.getCaratteriByIdAnimale(idAnimale);
        return list != null ? list : Collections.emptyList();
    }

    public List<TipoCarattere> getTipiCarattere() {
        List<TipoCarattere> tipi = caratteriDao.getTipoCarattere();
        if (tipi != null) {
            tipi.forEach(t -> t.setCaratteri(
                    caratteriDao.getCaratteriByTipo(t.getId_tipo_carattere())
            ));
        }
        return tipi != null ? tipi : Collections.emptyList();
    }

    @Transactional
    public String aggiungi(String idAnimale, Caratteri carattere) {
        carattere.setId_animale(idAnimale);
        carattere.setId_caratteri(null); // forza insert
        return caratteriDao.save(carattere);
    }

    @Transactional
    public void aggiornaNota(String idCaratteri, String idCarattere, String note) {
        Caratteri c = new Caratteri();
        c.setId_caratteri(idCaratteri);
        c.setId_carattere(idCarattere);
        c.setNote(note);
        caratteriDao.update(c);
    }

    @Transactional
    public void elimina(String idCaratteri) {
        caratteriDao.deleteByID(idCaratteri);
    }
}

File: frontend/src/app/services/animali/carattere.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

@Injectable({ providedIn: 'root' })
export class CarattereService {
  private apiUrl = environment.apiUrl + '/private/carattere';

  constructor(private http: HttpClient) {}

  getCaratteriAnimale(idAnimale: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/${idAnimale}`);
  }

  getTipiCarattere(): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/tipi`);
  }

  aggiungi(idAnimale: string, carattere: any): Observable<any> {
    return this.http.post(`${this.apiUrl}/${idAnimale}`, carattere);
  }

  aggiornaNota(idCaratteri: string, idCarattere: string, note: string): Observable<any> {
    return this.http.put(`${this.apiUrl}/${idCaratteri}/note`, { id_carattere: idCarattere, note });
  }

  elimina(idCaratteri: string): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${idCaratteri}`);
  }
}